### ISSUE TITLE
Getting started on a kind cluster

### ISSUE THREAD

**Issue by Str-Gen** [dask/dask-kubernetes#276]
Trying to run the basic example on a kubernetes-in-docker (kind) cluster has multiple issues. Two of which are most likely related to some missing networking setup. The third is a certificate issue.

### Local scheduler (deploy_mode="local") 
This does create the scheduler on the local machine and the pods on the cluster and it seems to link them up correctly, based on the internals of dask.distributed.Client. Trying to run a basic computation does run the computation on the workers in the cluster (seen from the resource usage in the dashboard), but then fails to aggregate on the host:

> distributed.scheduler - ERROR - Couldn't gather keys {"('mean_agg-aggregate-0f95a558ab91df8137659dd89dac2394',)": ['tcp://10.244.0.29:42993']} state: ['memory'] workers: ['tcp://10.244.0.29:42993']
> NoneType: None

After which it breaks the connection, reinstates the worker, tries again and fails again.

> distributed.scheduler - INFO - Remove worker <Worker 'tcp://10.244.0.29:42993', name: 1, memory: 1, processing: 0>
> distributed.core - INFO - Removing comms to tcp://10.244.0.29:42993
> distributed.scheduler - ERROR - Workers don't have promised key: ['tcp://10.244.0.29:42993'], ('mean_agg-aggregate-0f95a558ab91df8137659dd89dac2394',)
> NoneType: None
> distributed.client - WARNING - Couldn't gather 1 keys, rescheduling {"('mean_agg-aggregate-0f95a558ab91df8137659dd89dac2394',)": ('tcp://10.244.0.29:42993',)}
> distributed.scheduler - INFO - Register worker <Worker 'tcp://10.244.0.29:42993', name: 1, memory: 0, processing: 0>
> distributed.scheduler - INFO - Starting worker compute stream, tcp://10.244.0.29:42993
> distributed.core - INFO - Starting established connection
> 

### Remote scheduler (deploy_mode="remote")
Because there is some version mismatch between my local machine and the dask worker image and to try to avoid the potential network problem, I tried the remote (in-cluster) instantiation of the scheduler. This does create the scheduler as a pod on the kind cluster (single node) and the associated service:

> POD
> NAME                           READY   STATUS    RESTARTS   AGE    IP            NODE                 
> dask-dhgla-30d8a298-1hpbkc   1/1     Running   0          4m7s   10.244.0.31   kind-control-plane
> SERVICE
> NAME                      TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)             AGE
> dask-dhgla-30d8a298-1   ClusterIP   10.108.6.14   <none>        8786/TCP,8787/TCP   6m50s
> 

but it then can't seem to find the service which would point to a DNS issue or the same networking issue that plagues the comms between the local scheduler and the workers:

> Creating scheduler pod on cluster. This may take some time.
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/site-packages/distributed/comm/core.py", line 322, in connect
>     _raise(error)
>   File "/usr/lib/python3.8/site-packages/distributed/comm/core.py", line 275, in _raise
>     raise IOError(msg)
> OSError: Timed out trying to connect to 'tcp://dask-dhgla-30d8a298-1.default:8786' after 10 s: [Errno -2] Name or service not known

This fails very early and perhaps cleanup isn't called, because neither the pod, nor the service are removed from the cluster.

### Upon exit / termination -> 
While the others may be related to my setup, I'm more certain that this is a bug, but it creates a cascade of errors. It seems to repeatedly try to access the pod url of the workers. This happens even if you only create the scheduler (remote or local) and the workers, give them no work so they don't error and then tear them down upon exit. The workers that have been created do still clean up nicely though.

> Traceback (most recent call last):
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
>     httplib_response = self._make_request(
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 381, in _make_request
>     self._validate_conn(conn)
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
>     conn.connect()
>   File "/usr/lib/python3.8/site-packages/urllib3/connection.py", line 362, in connect
>     self.sock = ssl_wrap_socket(
>   File "/usr/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 352, in ssl_wrap_socket
>     raise SSLError(e)
> urllib3.exceptions.SSLError: ("unable to load trusted certificates: Error([('system library', 'fopen', 'No such file or directory'), ('BIO routines', 'BIO_new_file', 'no such file'), ('x509 certificate routines', 'X509_load_cert_crl_file', 'system lib')])",)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/weakref.py", line 642, in _exitfunc
>     f()
>   File "/usr/lib/python3.8/weakref.py", line 566, in __call__
>     return info.func(*info.args, **(info.kwargs or {}))
>   File "/home/dhgla/.local/lib/python3.8/site-packages/dask_kubernetes/core.py", line 695, in _cleanup_resources
>     pods = core_api.list_namespaced_pod(namespace, label_selector=format_labels(labels))
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/api/core_v1_api.py", line 15302, in list_namespaced_pod
>     return self.list_namespaced_pod_with_http_info(namespace, **kwargs)  # noqa: E501
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/api/core_v1_api.py", line 15413, in list_namespaced_pod_with_http_info
>     return self.api_client.call_api(
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 348, in call_api
>     return self.__call_api(resource_path, method,
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
>     response_data = self.request(
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/api_client.py", line 373, in request
>     return self.rest_client.GET(url,
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/rest.py", line 239, in GET
>     return self.request("GET", url,
>   File "/home/dhgla/.local/lib/python3.8/site-packages/kubernetes/client/rest.py", line 212, in request
>     r = self.pool_manager.request(method, url,
>   File "/usr/lib/python3.8/site-packages/urllib3/request.py", line 75, in request
>     return self.request_encode_url(
>   File "/usr/lib/python3.8/site-packages/urllib3/request.py", line 97, in request_encode_url
>     return self.urlopen(method, url, **extra_kw)
>   File "/usr/lib/python3.8/site-packages/urllib3/poolmanager.py", line 336, in urlopen
>     response = conn.urlopen(method, u.request_uri, **kw)
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 754, in urlopen
>     return self.urlopen(
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 754, in urlopen
>     return self.urlopen(
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 754, in urlopen
>     return self.urlopen(
>   File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 726, in urlopen
>     retries = retries.increment(
>   File "/usr/lib/python3.8/site-packages/urllib3/util/retry.py", line 439, in increment
>     raise MaxRetryError(_pool, url, error or ResponseError(cause))
> urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='127.0.0.1', port=40283): Max retries exceeded with url: /api/v1/namespaces/default/pods?labelSelector=foo%3Dbar%2Cbaz%3Dquux%2Cdask.org%2Fcluster-name%3Ddask-dhgla-18e6f52b-7%2Cdask.org%2Fcomponent%3Dworker%2Cuser%3Ddhgla%2Capp%3Ddask (Caused by SSLError(SSLError("unable to load trusted certificates: Error([('system library', 'fopen', 'No such file or directory'), ('BIO routines', 'BIO_new_file', 'no such file'), ('x509 certificate routines', 'X509_load_cert_crl_file', 'system lib')])")))

warning during execution:

> WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLError("unable to load trusted certificates: Error([('system library', 'fopen', 'No such file or directory'), ('BIO routines', 'BIO_new_file', 'no such file'), ('x509 certificate routines', 'X509_load_cert_crl_file', 'system lib')])"))': /api/v1/namespaces/default/pods?labelSelector=foo%3Dbar%2Cbaz%3Dquux%2Cdask.org%2Fcluster-name%3Ddask-dhgla-18e6f52b-7%2Cdask.org%2Fcomponent%3Dworker%2Cuser%3Ddhgla%2Capp%3Ddask

### Minimal Complete Verifiable Example:

Kind: nothing fancy, everything happens in the default context on a single node. No special roles, no RBAC. 

```bash
kind create cluster
```

The example from the docs with some extra code to check if workers get connected to the scheduler, which they do (to a local scheduler at least)

```python
from dask_kubernetes import KubeCluster
from dask.distributed import Client
from pprint import pprint
from time import sleep
import dask.array as da

cluster = KubeCluster(deploy_mode="local")
cluster.scale(2)  # specify number of workers explicitly

# Connect Dask to the cluster
client = Client(cluster)
sleep(15)
pprint(client.__dict__)
pprint(client.__dict__["_scheduler_identity"]["workers"])

# Create a large array and calculate the mean
array = da.ones((100, 100, 100))
print(array.mean().compute())  # Should print 1.0
```

### Environment:
Host
- Dask version: 2.26.0
- Dask_kubernetes version: 0.11.0
- Python version: 3.8.5
- Operating System: Arch Linux 5.8.13
- Install method (conda, pip, source): pip

Docker (worker and schedulers, remote deploy)
- Dask version: 2.30.0
- Python version: 3.8.0
- Operating System: Linux daskdev 5.8.13

### TL;DR: 
Trying to run the getting started example of kubernetes-docker on a standard kubernetes-in-docker (kind) cluster fails in various ways, both with a host-run and with a cluster-run scheduler. The issues are probably related to network setup that needs to be done before dask-kubernetes can be run on a kind cluster. However it is unclear what exactly is needed. The kind documentation isn't specific on allowing full network access either and the full Kubernetes docs may not apply to kind in particular.

**Comment by jacobtomlinson** [comment]
Thanks for raising this. I'm not sure we will be able to help though.

This project either assumes you are local to the kubernetes cluster (perhaps in a Jupyter session running in a pod) or remote to the cluster with standard service routing configured (Node ports or load balancers, etc).

Kind is more targeted at CI environments and testing. We use kind in the CI for this project, but run the tests from a pod within the cluster for exactly this reason.

If you are only intending on running Dask on a single machine then you probably don't need dask-kubernetes. And if you are running it on a multi-node kubernetes setup then you wont be using kind.

If you are intending to use kind to develop dask-kubernetes then I'm afraid I'll just need to refer you to the kind docs on networking. For development on linux I use k3s as it has load balancer support.
